Httpservices cachingcompiled new

jdeshin opened the pull request:
Предлагаемые изменения, в соответствии с описанным в форуме:
1. Введена опция advancedFeatures для:
throw; // Must fail!
_hostedScript.SetGlobalEnvironment(
2. Обработка запроса заканчивается End()
3. configPath перенесен из static конструктора, дабы его можно было подпихнуть и он подхватился после recycling хотя бы в винде.

jdeshin committed:
Неоконченный вариант. надо удалить из папки нагенеренные логи

jdeshin committed:
Мои предложения, по изложенному в форуме.
1. Обработка запроса оканчивается End()
2. Введена опция advancedFeatures для:
_hostedScript.SetGlobalEnvironment
throw; // Must fail!
3. configPath перенесен из static в конструктор экземпляра, чтоб на "лету" его можно было менять в win по крайней мере

jdeshin committed:
Для throw //Must fail и подобных штук
Выделена отдельная опция:
errorActionPreferenceWhenHandlerCreating
Если == stop - завершаемся с исключением, если нет - продолжаем

EvilBeaver commented on src/ASPNETHandler/ASPNETHandler.cs:
```diff
@@ -56,6 +53,8 @@ static ASPNETHandler()
                     }
                     // TODO: Исправить - должно падать. Если конфиг сайта неработоспособен - сайт не должен быть работоспособен.
+                    // Спорное утверждение. Это загрузка внешних компонент aka sql etc. 
+                    // И что из-за одной dll, которой нет физически, она не используется и она есть в конфиге огрести неработоспособное приложение? 
```
> из-за одной dll, которой нет физически, она не используется и она есть в конфиге огрести неработоспособное приложение

Блин, ну разумеется! Как минимум предупреждение выдать куда-то, но не совсем заигнорить!

jdeshin replied:
Лог?

jdeshin replied:
Сделал логгирование, аналогично тому, как в конструкторе

EvilBeaver commented on src/ASPNETHandler/ASPNETHandler.cs:
```diff
@@ -143,30 +142,36 @@ public ASPNETHandler()
-                _hostedScript.Initialize();
+                // давай использовать эту опцию для разных вот таких недоделанных случаев
+                // потом будем или удалять или менять название опции, если это опционально
+                if (appSettings["advancedFeatures"] == "true")
```
Предлагаю другое название. HandlerLoadingPolicy: {strict|tolerant}, где tolerant игнорирует ошибки (твой вариант), а strict выдает их (мой).

EvilBeaver commented on src/ASPNETHandler/ASPNETHandler.cs:
```diff
@@ -143,30 +142,36 @@ public ASPNETHandler()
-            WriteToLog(logWriter, "End loading.");
-            // Закрываем лог, если надо
-            if (logWriter != null)
+                if (appSettings["errorActionPreferenceWhenHandlerCreating"] == "stop")
```
повесить на ту же опцию LoadingPolicy?

jdeshin replied:
Тут вроде как это разные вещи, расширенные фичи и политика загрузки.
Мож мне фичи то нужны, а загрузку хочу tolerant

jdeshin replied:
И мож таки stop/continue. Более привычны неанглоязычному человеку :)

jdeshin replied:
Заменил appSettings["errorActionPreferenceWhenHandlerCreating"]  на HandlerLoadingPolicy: {strict
Добавил при всех загрузках. Так норм?

EvilBeaver commented on src/ASPNETHandler/ASPNETHandler.cs:
```diff
@@ -182,40 +187,46 @@ void WriteToLog(TextWriter logWriter, string message)
-                module = cache[context.Request.PhysicalPath] as LoadedModuleHandle?;
+                #region Загружаем скрипт (файл .os)
+                // Кэшируем исходный файл, если файл изменился (изменили скрипт .os) загружаем заново
+                // В Linux под Mono не работает подписка на изменение файла.
```
Предлагаю вынести в отдельный метод все, что в блоке try, чтобы код обработчика был более явным:

```c#
try
{
   callScriptHandler(context) 
}
finally
{
   context.Response.End;
}
```

jdeshin replied:
Готово, назвал CallScriptHandler

EvilBeaver reviewed (changes requested):
/cc @dmpas @nixel2007

jdeshin committed:
Добавил везде:
if (appSettings["handlerLoadingPolicy"] == "strict")
                            throw; // Must fail!

EvilBeaver commented on src/ASPNETHandler/ASPNETHandler.cs:
```diff
@@ -56,39 +58,33 @@ static ASPNETHandler()
-                    catch {/*не загрузилась, ничего не делаем*/ }
+                    catch(Exception ex)
+                    {
+                        WriteToLog(logWriter, "Error loading assembly: " + assemblyName + " " + ex.Message);
```
Вместо e.Message правильно логировать e.ToString()

К слову, С. Тепляков предлагал расстреливать тех, кто так не делает.

jdeshin replied:
Это его личные так сказать трудности :), поправлю

jdeshin replied:
Поправил, обновил

EvilBeaver commented on src/ASPNETHandler/ASPNETHandler.cs:
```diff
@@ -138,38 +134,60 @@ public ASPNETHandler()
-                _hostedScript.SetGlobalEnvironment(new ASPNetApplicationHost(), new AspEntryScriptSrc());
-                _hostedScript.Initialize();
+                if (appSettings["advancedFeatures"] == "true")
+                    _hostedScript.SetGlobalEnvironment(new ASPNetApplicationHost(), new AspEntryScriptSrc());
```
Это условие не нужно. Без установки среды не будут работать методы Сообщить и половина библиотек

EvilBeaver replied:
Т.е. должно быть безусловная установка, без опции, к.м.к

jdeshin replied:
Ну давай по третьему кругу :)
Что конкретно не будет работать? Исходя из моих тестов, страницы процессятся, библиотеки загружаются. Пожалуйста поясни, что не будет работать и зачем в веб-приложении Console.Writeline. На сколько я помню, ты говорил, что хост нужен для вывода Сообщить. Нам некуда чего-либо сообщать ибо это не консольное приложение. В 1С такая же ситуация на сервере. Ну таки что не будет работать? Таки нужно понимание.

jdeshin replied:
Сделал и добавил дефолтные заглушки для host и исходного скрипта. Ключ в силе. У тебя там оказывается много интересного в глобальном контексте :) Действительно, если ты просто переносишь консольные скрипты - может отваливаться. По моему все, и волки сыты и овцы целы

EvilBeaver replied:
@jdeshin я использую библиотеку логирования logos. Я же имею право это делать, как автор хттп-сервиса? Ну так вот, я хочу, чтобы вызовы моего сервиса логировались. Однако logos не работает, т.к. вызывает метод СтартовыйСценарий(). Кроме того, я не против, если в консоль моего xsp4 будет также выводиться лог. Я этот лог потом прочту, например, через docker logs. Вполне себе допустимое действие, однако в Консоль писать через Сообщить не могу, т.к. вместо хоста у меня заглушка.

Итого, просьба оставить на месте вызов метода SetGlobalEnvironment.

jdeshin committed:
Поправил e.Message на e.ToString()

jdeshin committed:
Добавил заглушки по умолчанию для хоста и стартового скрипта

jdeshin committed:
Поменял advancedFeatures на classicHTTPService.
Если опция есть и true, то используются пустые заглушки. По умолчанию используются твои

jdeshin committed:
Поправил в соответствии с sonar

jdeshin committed:
Поправил в соответствии с sonar

jdeshin committed:
mm

EvilBeaver commented on src/ASPNETHandler/ASPNETHandler.cs:
```diff
@@ -138,38 +134,62 @@ public ASPNETHandler()
                 // метод настраивает внутренние переменные у SystemGlobalContext
-                _hostedScript.SetGlobalEnvironment(new ASPNetApplicationHost(), new AspEntryScriptSrc());
-                _hostedScript.Initialize();
+                if (appSettings["classicHTTPService"] == "true")
```
Зачем вот эти сложности и параметры? Что произойдет, если этого параметра вообще не будет, а SetGlobalEnvironment будет вызываться безусловно?

jdeshin replied:
Да ничего не произойдет, если класс Host будет типа заглушкой и поведение не будет отличаться от поведения платформы. У тебя же там console.writeline, цвета какие-то текста. Я, как лицо с ограниченными интеллектуальными возможностями - ниасилил. Я то не против, если это будет работать быстро, понятно и прозрачно. Ведь ты посмотрел заглушки и думаю - у тебя не возникло трудностей с пониманием. Я уже посмотрел и понял, что сообщить будет валиться с исключением, да, это неправильно и не соответствует поведению платформы. (Мог бы и объяснить :)) Там еще есть входные параметры - которых у нас нет, опять таки в силу того, что это не консольный скрипт. По хорошему, эти вопросы с хостом надо решать в движке, мало ли кто его куда захочет воткнуть. Но так быстрее и проще. Объясни плз. что ты собираешься делать в этих классах и не будет ли тормозить.
Надеюсь на понимание :)

EvilBeaver replied:
Тормозить не будет, а эти классы исключительно для штатных библиотек 1Script, которые активно юзают стартовый скрипт и Сообщить

EvilBeaver commented on src/ASPNETHandler/ASPNETHandler.cs:
```diff
@@ -138,38 +134,62 @@ public ASPNETHandler()
-                _hostedScript.SetGlobalEnvironment(new ASPNetApplicationHost(), new AspEntryScriptSrc());
-                _hostedScript.Initialize();
+                if (appSettings["classicHTTPService"] == "true")
+                    _hostedScript.SetGlobalEnvironment(new AspNetNullApplicationHost(), new AspNetNullEntryScriptSrc());
```
NullEntryScript - это самообман, т.к. EntryScript существует - это сам файл os нашего сервиса. Стало быть, неправильно заменять результат этого метода на заглушку. Зачем? Чтобы что?

jdeshin replied:
ОК, тогда поясни плз. Это - конструктор хэндлера, он вызывается в общем случае до каких либо запросов. Формируется пул из нескольких экземпляров. При этом никаких скриптов и кода не выполняется. Что тут должно быть.
Случай номер два - один и тот-же хэндлер обрабатывает запросы к разным файлам os, Тогда его надо устанавливать как-то в ProcessRequest. Но ты его обозначил в конструкторе. Его можно перенести без влияния на инициализацию? И что туда подставлять?

EvilBeaver replied:
> один и тот-же хэндлер обрабатывает запросы к разным файлам os

Все, теперь вспомнил. Эта проблема не решается в текущем состоянии хостед. Поэтому, этот мой класс указывает на web.config

https://github.com/EvilBeaver/OneScript/blob/develop/src/ASPNETHandler/AspEntryScriptSrc.cs#L21

Это тот еще костыль, просто если стартовый скрипт будет указывать на несуществующий файл, то библиотеки поотваливаются многие (logos - точно). Они ожидают там файл.

jdeshin replied:
А web.config в общем случае может и не быть. Он может быть в рутовой папке, более "локальный" web.config переопределяет параметры более рутового.
Принципиально какой файл? Что они с ним делают?

jdeshin replied:
Вот поэтому на мой взгляд - опции рулят :) Понятно, что есть куча унаследованного кода, который адаптировать крайне сложно и он должен поддерживаться. С другой стороны - чисто веб-приложениям это не надо и я бы сказал, на мой взгляд,  в какой-то мере методологически вредно (ну организовывать логи через Сообщить к примеру) :). Так фигак - выставил опцию supportConsoleApp к примеру и все норм, не выставил - обычное веб приложение без требованиям к наличию файлов. Подумай, этож нормально и есть в том же 1С режим совместимости.

EvilBeaver replied:
ОО! Идея. Давай вместо туманной опции classicHTTPService введем две другие: startupScript и enableEcho

Первая - что написали, то и будет возвращено в "Стартовый сценарий". Ничего не написали - будет NullScript

Вторая - включен Сообщить - пишем в консоль. Выключен - Null-объект
Тормозить не будет.

jdeshin replied:
Давай :). Тогда наверное прими pr там логи и политика загрузки и добавь опции как надо.

artbear commented:
На Ваш тред даже Гитхаб уже ругается :)


[image: Inline images 1]

PS упс, я еще чуток длину увеличил :)

С уважением,
Артур Аюханов

2018-01-11 17:59 GMT+03:00 jdeshin <notifications@github.com>:

> *@jdeshin* commented on this pull request.
> ------------------------------
>
> In src/ASPNETHandler/ASPNETHandler.cs
> <https://github.com/EvilBeaver/OneScript/pull/655#discussion_r160981944>:
>
> >                          }
>                      }
>                  }
>
>                  // метод настраивает внутренние переменные у SystemGlobalContext
> -                _hostedScript.SetGlobalEnvironment(new ASPNetApplicationHost(), new AspEntryScriptSrc());
> -                _hostedScript.Initialize();
> +                if (appSettings["classicHTTPService"] == "true")
> +                    _hostedScript.SetGlobalEnvironment(new AspNetNullApplicationHost(), new AspNetNullEntryScriptSrc());
>
> Вот поэтому на мой взгляд - опции рулят :) Понятно, что есть куча
> унаследованного кода, который адаптировать крайне сложно и он должен
> поддерживаться. С другой стороны - чисто веб-приложениям это не надо и я бы
> сказал, на мой взгляд, в какой-то мере методологически вредно (ну
> организовывать логи через Сообщить к примеру) :). Так фигак - выставил
> опцию supportConsoleApp к примеру и все норм, не выставил - обычное веб
> приложение без требованиям к наличию файлов. Подумай, этож нормально и есть
> в том же 1С режим совместимости.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/EvilBeaver/OneScript/pull/655#discussion_r160981944>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACyRcTcULk4X4_vRGHV7HpgWhqlahZRKks5tJiHFgaJpZM4RZHvg>
> .
>
